common/power_button.c

```diff
@@ -115,6 +115,624 @@ static void power_button_init(void)
 }
 DECLARE_HOOK(HOOK_INIT, power_button_init, HOOK_PRIO_INIT_POWER_BUTTON);
 
+/*
+ *====================
+ * After G3 State
+ *====================
+ *
+ * Required headers:
+ *   #include "chipset.h"
+ *   #include "common.h"
+ *   #include "console.h"
+ *   #include "hooks.h"
+ *   #include "host_command.h"
+ *   #include "system.h"
+ *   #include "util.h"
+ *
+ * Only useful for boards with dedicated 3V coin battery to sustain bbram, and
+ * only for EC images compiled with CONFIG_POWER_BUTTON_INIT_IDLE, as this
+ * ensures that OFF bbram reset flags are retained.
+ *
+ * Available "board/NAME/board.h" config macros:
+ *
+ *==========
+ *== Enable After G3 State, otherwise legacy OFF flag handling will be used.
+ *==========
+ * #define CONFIG_AFTER_G3_STATE
+ *
+ *==========
+ *== Use RESET_FLAG_AP_OFF/EC_RESET_FLAG_AP_OFF as OFF flag, instead of using
+ *== EC_RESET_FLAG_AP_IDLE as OFF flag.
+ *==
+ *== FIZZ, and older boards, do not support AP_IDLE, and should use AP_OFF as
+ *== OFF flag. PUFF, and newer boards, have been migrated to AP_IDLE as OFF
+ *== flag, and should use AP_IDLE.
+ *==========
+ * #define CONFIG_AFTER_G3_STATE_USE_AP_OFF_AS_OFF_FLAG
+ *
+ *==========
+ *== Enables fix for bbram bug in older EC code, where presence of *any* bbram
+ *== reset flags after a jump will discard jump data, and not mark image state
+ *== as jumped to.
+ *==
+ *== Bug is caused by the fact that when EC image is executed after a jump, it
+ *== first restores system reset flags from bbram reset flags via pathway:
+ *==   main()->gpio_pre_init()->system_is_reboot_warm()->
+ *==     system_check_reset_cause()
+ *== Then it overwrites system reset flags with jump data reset flags via
+ *== pathway:
+ *==   main()->system_common_pre_init()
+ *==
+ *== In bugged firmware system_common_pre_init() will only restore jump data if
+ *== condition "system_reset == 0" (system reset flags == 0) evaluates to TRUE.
+ *==
+ *== This creates a risk of causing auto power on if device is off, or risk of
+ *== inducing abnormal state (initializing power button as off) if device is
+ *== running.
+ *==
+ *== Upstream fixed this is newer EC code by dropping the problematic condition:
+ *==   https://github.com/MrChromebox/chrome-ec/commit/e833bdec81b9
+ *== However, this is mainly suitable for fixing RW (jump RO->RW), as fixing RO
+ *== would require overwriting factory RO, which poses a risk of bricking
+ *== device, with no easy way to recover.
+ *==
+ *== Fixing RO (jump RW->RO), without actually changing RO, has two pathways:
```

Comment:
I'm not sure this is really a path we need to worry about, given the complexity of the implementation